src/lib/types/Discount.ts

```diff
@@ -11,4 +11,6 @@ export interface Discount extends Timestamps {
 	wholeCatalog: boolean;
 	beginsAt: Date;
 	endsAt: Date | null;
+	mode?: 'productPercentage' | 'freeProductQuantity';
+	quantityPerProduct?: Record<string, number>;
 }
```

Comment:
```suggestion
} & ({
	mode: 'percentage';
	percentage: number;
} | {
	mode: 'freeProducts';
	quantityPerProduct: Record<string, number>;
})
```

and a migration to add `mode` to existing discounts

you can use `z.union` on just the two fields when you need to use zod / clean up typing